### ISSUE TITLE
Faceting doesn't work correctly--on several fronts

### ISSUE THREAD

**Issue by nrshapiro** [Psychoanalytic-Electronic-Publishing/PEP-Web-User-Interface#518]
Faceting no longer works correctly (as it did in most cases at least in April).  

1) After a search, selecting a single facet should reduce the result set, it's being intersected with the results.
2) Adding a second facet should then enlarge the results, since the intersection of the search is to the union of the facets.

We have a somewhere between the client and server in how the queries are being sent (by the client) or handled (by the server).  Likely both.  I had put in some workaround/optimizations to appropriately handle how the client sent more complex queries (particularly author), but for that field, I believe those are making things worse.  (For example: accepting author first name last name and lastname firstname and lastname first initial.)

(This is one reason I need to have the local client working...I can't even see the problem well.  But I think we'll also need to work together on this one.)

**Comment by nrshapiro** [comment]
@abaker-gavant @adistasio 

Ok, this may only be step one, but it's clearly a problem.  

     "Tuckett, David" AND (taffler, richard OR amati mehler, jacqueline)

In this author query, the names in () on the right are from the facets.  Since they are more than one word, they need to be in quotes, like "Tuckett, David"

This one was hard to catch, because in many cases, like this one, the intersection of the two limits the results and hides the actual search problem with the facets.

I note that some of the other multiword facets, like the journal names, are already coming in with quotes.

In Fulltext1: 
    art_sourcetitleabbr:("Int. J. Psychoanal." OR "Int. Rev. Psycho-Anal." OR "Brit. J. Psychother.")

Looking at how well this works, I may ask you to similarly prefix other facets with the facet field.  

The problem with author, for example

Tuckett, David" AND (taffler, richard OR amati mehler, jacqueline)

becomes more difficult for me to handle when the search part, "Tuckett, Daivd", needs to be searched in three different fields, because it can be firstname then last, last than initial, or a standard id!  Currently, I'm doing both parts in multiple fields, but the facets really only need to be searched in the field from which they came. 

The author field is somewhat of a pattern, which I can parse, but it can become more complicated if there's an AND used in the query.  So I am proposing to do the following:

I will add a facetquery parameter to the search endpoint.  You would have to use it like you do Fulltext1 for journals above, using the name of the facet field and then the query in parens, with quotes where necessary.

That way the query fields stay pure, and I can set them against multiple fields as needed, and the facets come with the correct search field indicated for each.  E.g., if both author and journal facets were used, you would submit:

In parameter facetquery:

   art_sourcetitleabbr:("Int. J. Psychoanal." OR "Int. Rev. Psycho-Anal." OR "Brit. J. Psychother.") AND art_authors:("taffler, richard" OR "amati mehler, jacqueline")

Here you can see that this works very well and then can easily be combined with any other search field of the form (I've simulated the facets area using adv:: search in smartSearch):

![image](https://user-images.githubusercontent.com/4837554/119164323-4b522700-ba2a-11eb-8ef0-815424a82b3b.png)

I should be able to get the field in very quickly...and will post to stage for your adoption, unless you chime in with issues for this plan.

**Comment by nrshapiro** [comment]
@abaker-gavant @adistasio 

Ok, I've added a parameter facetquery to the server on Stage.  It should still run the old code as before, but I think we should move all the facets queries there.  Since the client specifies the return facet fields when doing a query, it also always knows the exact field(s) to use in facetquery.  The facetquery parameter is simply ANDed to whatever ends up in the solr fq parameter.  There's no need for the server to interpret or parse, at least as as I see now.

I've tested facetquery in search, ran the whole test suite, but I'm sure there are other endpoints like Analysis that are not getting a test with the new parameter, so that will be my next task, but meanwhile, you have something you can work with.

**If it's possible to make the changes, at least to author searches today, that would be great.**  The way that works currently is the biggest issue.  That will also mean I can do more testing on the weekend and see how it works.

**Comment by nrshapiro** [comment]
@bakerac4 (Adam, see above, sorry, I was using your other address, not sure if you got those notifications, but I edited in the usual one and writing this new one to be sure you've seen.)

**Comment by bakerac4** [comment]
@nrshapiro I commited my work in `story/gh518-search-facets`. I dont think its exactly what you want but its going to take some time to parse through the changes I have to make as the search/search facet stuff is fairly complex.

I tried to reproduce the query you were using and got fairly close. I added `art_sourcetitleabbr:("Int. J. Psychoanal." OR "Int. Rev. Psycho-Anal." OR "Brit. J. Psychother.")` in the smart search text box and then searched for Amati Mehler in the author box. Then In order to get the facetquery param to show up I clicked on the facet of "Amati mehler, jacqueline" which ends up giving me this

```abstract=true&author=%22Amati+Mehler%22&facetfields=art_year_int%2Cart_views_last12mos%2Cart_cited_5%2Cart_authors%2Cart_lang%2Cart_type%2Cart_sourcetype%2Cart_sourcetitleabbr%2Cglossary_group_terms%2Cart_kwds_str&facetlimit=15&facetmincount=1&facetquery=art_authors%3A(%22amati+mehler%2C+jacqueline%22)&formatrequested=XML&highlightlimit=5&limit=20&offset=0&smarttext=art_sourcetitleabbr%3A(%22Int.+J.+Psychoanal.%22+OR+%22Int.+Rev.+Psycho-Anal.%22+OR+%22Brit.+J.+Psychother.%22)&sort=bibliographic&synonyms=false&user=true```. Obviously not exactly what your expecting though..

Also Im a bit confused at why your saying 

```art_sourcetitleabbr:("Int. J. Psychoanal." OR "Int. Rev. Psycho-Anal." OR "Brit. J. Psychother.") AND art_authors:("taffler, richard" OR "amati mehler, jacqueline")``` should be correct ant not just ```art_authors:("taffler, richard" OR "amati mehler, jacqueline")``` based on your screenshot. The only facets you have selected are "Amati mehler, jacqueline"

Ill pick this up first thing monday morning

**Comment by nrshapiro** [comment]
@bakerac4 - Adam, I just rebooted and found my email app wasn't picking up/downloading a lot of emails, including notifications...so I didn't see this till now.  Will look at it in a bit, but right now going through all the missed events!

**Comment by nrshapiro** [comment]
@bakerac4 The screenshots were me showing you a simulation of what the facetquery field would receive, before I implemented it.  No facets were actually used or harmed in that simulation. 

I simulated what was needed using adv query syntax in SmartSearch.  That lets you do a multiple field query.  Without that, you can search a field, but only one.

After doing that, and writing that, I implemented the facetquery parameter, which was then pushed up on Stage.

**Comment by bakerac4** [comment]
@nrshapiro ok, can you confirm whether authors are working the way you would like then? If so I can work on adding in the rest now?

**Comment by nrshapiro** [comment]
@bakerac4 I may be wrong, but my understanding is that I can't build the client and test from your "story" you posted, you need to post a full build with it incorporated.  At least I don't know how to do that.

![image](https://user-images.githubusercontent.com/4837554/119347278-50e98000-bc69-11eb-857c-9d5fc2424418.png)

**Comment by bakerac4** [comment]
@nrshapiro now that you have dev working you should be able to check out that branch in GIT and run it locally

**Comment by nrshapiro** [comment]
@bakerac4 Thanks--I misunderstood what Alan had said about the Stories a while ago and thought they were partial rather than complete builds.  I was able to install just fine.  

However, what you are doing is not per the new method.  Interestingly, it works better than before, but it doesn't solve my problem where I need to process fields like Author in a "special" manner, and thus I need facets to be presented separately.  For example, with a query like Tuckett in author and selecting a few facets, they still seem to be coming in on the author field.

There's a new parameter you can see per my example above from yesterday.  It's called facetquery.  That's where ALL the facets go.  They can "all" go there (for any field) because they need to use syntax naming the facet field they came from.  See my example in the interactive interface above.  

Here's a different one (simpler still):
`https://stage-api.pep-web.rocks/v2/Database/Search/?synonyms=false&facetquery=art_authors%3A%22bournova%2C%20klio%22&author=%22B%C3%A9atrice%20Ithier%22&viewperiod=4&abstract=false&formatrequested=HTML&similarcount=0&highlightlimit=5&facetmincount=1&facetlimit=15&facetoffset=0&sort=score%20desc&limit=15&offset=0`

url decoded: 
`https://stage-api.pep-web.rocks/v2/Database/Search/?synonyms=false&facetquery=art_authors:"bournova, klio"&author="Béatrice Ithier"&viewperiod=4&abstract=false&formatrequested=HTML&similarcount=0&highlightlimit=5&facetmincount=1&facetlimit=15&facetoffset=0&sort=score desc&limit=15&offset=0`

Visual:

![image](https://user-images.githubusercontent.com/4837554/119352253-72e60100-bc6f-11eb-9090-062e943b49ea.png)

Perhaps I've loaded something wrong and you actually did this in the Story, but I'm not seeing it.  If you want to call me--I think you already have my number, or @adistasio does--we can chat about this to make sure we're on the same page before you do any more conversions.

**Comment by bakerac4** [comment]
@nrshapiro for our call today Im hoping you could give me a few use cases that I can try and reproduce through the actual application. That would help out alot!

**Comment by nrshapiro** [comment]
Use Case: 
```
        encodedarg = urllib.parse.quote_plus("David Tuckett OR Peter Fonagy")
        encodedarg2 = urllib.parse.quote_plus('art_sourcetitleabbr:("Int. J. Psychoanal." OR "Int. Rev. Psycho-Anal." OR "Brit. J. Psychother.") AND art_authors:("taffler, richard" OR "amati mehler, jacqueline")')
        full_URL = base_plus_endpoint_encoded(f"/v2/Database/Search/?author={encodedarg}&facetquery={encodedarg2}")
        print (f"Full URL: {full_URL}")
        response = requests.get(full_URL, headers=headers)
        assert(response.ok == True)
        r = response.json()
        print (r)
        response_info = r["documentList"]["responseInfo"]
        response_set = r["documentList"]["responseSet"] 
        assert(response_info["fullCount"] == 2)


```

**Comment by nrshapiro** [comment]
The new method works great for authors.

**Comment by nrshapiro** [comment]
@bakerac4  Alas, I had only checked authors...tonight I tested the others, and found problems.  Perhaps you weren't done, but I'll present all the details here to help just in case.

There rules for the new parameter facetquery are basically to use Solr syntax.  There is currently no processing on this field, because of the number of possibilities and a string of query conditions this field can contain.

First, let's start with the domain of facets you are currently using (any field could be used, which is why I am leaving this up to the client to do this part).  The Solr facets being used include: 

`['art_year_int', 'art_views_last12mos', 'art_cited_5', 'art_authors', 'art_lang', 'art_type', 'art_sourcetype', 'art_sourcetitleabbr', 'glossary_group_terms', 'art_kwds_str']
`

When you build the query for the facetquery parameter, its' no different than you do in the fulltextfield1 parameter.  Basically, the rules are:

1) Each facetquery must be preceded by the field used for that facet and then a colon.
2) The query presented after the colon needs to use Solr syntax for ranges.  A range in Solr is in Square brackets.
3) The facetfields and the query  can be anded and or'd together, using && and || or AND and OR per solr  syntax.  That includes ranges (see example below)
4) Field values with more than one word shall be in ""

Examples that would work in the OPAS facetquery parameter:

```
art_type:(TOC || Review || Announcement)
art_sourcetype:(Book || Journal)
art_cited_5:([0 TO 9] || [10 TO 20])
art_year_int:([1880 TO 1899] || [1920 TO 1930])
```

And of course, you can combine more than one facet in the facetquery parameter:

`
facetquery=art_type:(TOC || Review || Announcement) && art_sourcetype:(Book || Journal) && art_cited_5:([0 TO 9] || [10 TO 20])
`

**Incorrect Usage Seen...Actual examples from APP:**
- art_type, missing the field name
- art_sourcetype missing the field name
- art_cited_5 comes in missing the field name and with incorrect query syntax for Solr.  It comes in as (0 TO 9 IN 5).  The IN 5 is meaningless here and Solr doesn't understand it.  It should just be art_cited_5 and the query as shown in my examples.
- art_lang,  missing the field name
- art_year_int, missing the field name, needs to use Solr ranges in square brackets, e.g., [1880 TO 1900]
- art_views_last12mos, , missing the field name

**Correct Usage Seen...Actual examples from APP:**
- art_authors:("freud")
- art_sourcetitleabbr:("Int. Psycho-Anal. Lib.")
- glossary_group_terms:(dreaming)
- art_kwds_str:("perversion. homosexualité. patricide. matricide." OR "gespaltene loyalität")

**Comment by bakerac4** [comment]
@nrshapiro I believe I have the cases you mentioned all fixed. Give it a try and let me know. If everything works Ill merge this into master

**Comment by bakerac4** [comment]
Fixed by #525
